### DISCUSSION TITLE
tutorials/math/random_number_generation

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#74] [User-contributed notes]
# tutorials/math/random_number_generation

Many games rely on randomness to implement core game mechanics. This page guides you through common types of randomness and how to implement them in Godot. After giving you a brief overview of usef...

https://docs.godotengine.org/en/stable/tutorials/math/random_number_generation.html

**Comment by IanMancuso** [comment]
In the "Get A Random Array Element" section, could there be a reference to the pick_random() method of arrays?

**Reply by Calinou** [reply]
Indeed. `Array.pick_random()` was added after this page was written, so that's why it's not mentioned right now.

I've opened a PR to mention it: https://github.com/godotengine/godot-docs/pull/9822

**Comment by Bloodyaugust** [comment]
Does an instance of `RandomNumberGenerator` with seed `x` and godot version `y` generate the same sequence of random numbers across OS/platform/hardware?

**Reply by Calinou** [reply]
As far as I know, it *should* be consistent across OS/platform/hardware, but it probably can't be 100% guaranteed.

**Comment by riquardbeks** [comment]
So this is pretty interesting. I am still learning but whatever i do within the editor i always get the same results. Sort of..

```gdscript
extends Node2D


func _ready() -> void:

	var loop_number = 1000000
	var number0 :int = 0
	var number1 :int = 0
	var number2 :int = 0
	var number3 :int = 0
	var number4 :int = 0
	var number5 :int = 0
	var number6 :int = 0
	var number7 :int = 0
	var number8 :int = 0
	var number9 :int = 0
	var number10 :int = 0

	while loop_number >= 0:
		
		var randomnumber: RandomGeneration = RandomGeneration.new()
		#print (randomnumber._generate_random_number())

		if randomnumber._generate_random_number() == 0:
			number0 +=1
		elif randomnumber._generate_random_number() == 1:
			number1 += 1
		elif randomnumber._generate_random_number() == 2:
			number2 += 1
		elif randomnumber._generate_random_number() == 3:
			number3 += 1
		elif randomnumber._generate_random_number() == 4:
			number4 += 1
		elif randomnumber._generate_random_number() == 5:
			number5 += 1
		elif randomnumber._generate_random_number() == 6:
			number6 += 1
		elif randomnumber._generate_random_number() == 7:
			number7 += 1
		elif randomnumber._generate_random_number() == 8:
			number8 += 1
		elif randomnumber._generate_random_number() == 9:
			number9 += 1
		elif randomnumber._generate_random_number() == 10:
			number10 += 1

		if loop_number == 0:
			print("Number 0 was printed :","\t", number0," times")
			print("Number 1 was printed :","\t", number1," times")
			print("Number 2 was printed :","\t", number2," times")
			print("Number 3 was printed :","\t", number3," times")
			print("Number 4 was printed :","\t", number4," times")
			print("Number 5 was printed :","\t", number5," times")
			print("Number 6 was printed :","\t", number6," times")
			print("Number 7 was printed :","\t", number7," times")
			print("Number 8 was printed :","\t", number8," times")
			print("Number 9 was printed :","\t", number9," times")
			print("Number 10 was printed :","\t", number10," times")
		
		loop_number -= 1


class RandomGeneration:
	func _generate_random_number():
		return randi_range(0,10)
```


This is the output and it always like this. Try it out. How is this happening?

```
Godot Engine v4.3.stable.official.77dcf97d8 - https://godotengine.org
OpenGL API 3.3.0 NVIDIA 561.09 - Compatibility - Using Device: NVIDIA - NVIDIA GeForce RTX 4070

Number 0 was printed :	90941 times
Number 1 was printed :	82972 times
Number 2 was printed :	74985 times
Number 3 was printed :	68529 times
Number 4 was printed :	62361 times
Number 5 was printed :	56447 times
Number 6 was printed :	51591 times
Number 7 was printed :	46281 times
Number 8 was printed :	42265 times
Number 9 was printed :	38621 times
Number 10 was printed :	34852 times

--- Debugging process stopped ---
```

**Reply by riquardbeks** [reply]
Even more funny, it never loops to the number i give in. When I did 100 000 it only did 65.392 loops

**Reply by Kiokurashi** [reply]
The reason for this distribution and for the number of executions being less than your loop are probably one and the same: You're generating a number each time you do a check in the if condition (or at least I think that's how that works, I'm not sure since you recreate the randomnumber variable each loop) so it is giving you an increase each time it hits one of those (which then exits the if block) so the only ones that can be 10 is if the RNG fails on 0-9 AND succeeds on a random 10 (if it works the way I think it does then that's not a guaranteed result). Otherwise you get no increase, but still increase the loop count.

Also, I find it interesting that you chose to use 11 variables and if-else statements instead of a for loop that takes the generated number as the index of an array to add the value to. I don't know enough to know if one is better than the other.

**Reply by Reynaeris** [reply]
I pulled it into my own project to puzzle it out. It's actually because of the way your if statements are setup. Each if block is generating a new number between 0 and 10 and seeing if it equals the number in its branch. In a fair number of your while loops, none of the if statements have this condition met. So it's running all the times you're telling it to, but sometimes it gets to the end of the statements without incrementing anything. 

To fix it you need to pull the number generations out of the if statements and do it at the top of the while loop instead. Then compare the single generated number to all the possible numbers it could be. 

```gdscript
extends Node2D


func _ready() -> void:

	var loop_number = 1000000
	var number0 :int = 0
	var number1 :int = 0
	var number2 :int = 0
	var number3 :int = 0
	var number4 :int = 0
	var number5 :int = 0
	var number6 :int = 0
	var number7 :int = 0
	var number8 :int = 0
	var number9 :int = 0
	var number10 :int = 0

	var randomnumber: RandomGeneration = RandomGeneration.new()
	while loop_number >= 0:
		var rand_num = randomnumber._generate_random_number()
		#print (randomnumber._generate_random_number())

		if rand_num == 0:
			number0 +=1
		elif rand_num == 1:
			number1 += 1
		elif rand_num == 2:
			number2 += 1
		elif rand_num == 3:
			number3 += 1
		elif rand_num == 4:
			number4 += 1
		elif rand_num == 5:
			number5 += 1
		elif rand_num == 6:
			number6 += 1
		elif rand_num == 7:
			number7 += 1
		elif rand_num == 8:
			number8 += 1
		elif rand_num == 9:
			number9 += 1
		elif rand_num == 10:
			number10 += 1

		if loop_number == 0:
			print("Number 0 was printed :","\t", number0," times")
			print("Number 1 was printed :","\t", number1," times")
			print("Number 2 was printed :","\t", number2," times")
			print("Number 3 was printed :","\t", number3," times")
			print("Number 4 was printed :","\t", number4," times")
			print("Number 5 was printed :","\t", number5," times")
			print("Number 6 was printed :","\t", number6," times")
			print("Number 7 was printed :","\t", number7," times")
			print("Number 8 was printed :","\t", number8," times")
			print("Number 9 was printed :","\t", number9," times")
			print("Number 10 was printed :","\t", number10," times")
		
		loop_number -= 1


class RandomGeneration:
	func _generate_random_number():
		return randi_range(0,10)
```
And the results:
```
Godot Engine v4.3.stable.arch_linux - https://godotengine.org
OpenGL API 3.3.0 NVIDIA 565.57.01 - Compatibility - Using Device: NVIDIA - NVIDIA GeForce RTX 2060

Number 0 was printed :	90682 times
Number 1 was printed :	91043 times
Number 2 was printed :	90854 times
Number 3 was printed :	91045 times
Number 4 was printed :	91220 times
Number 5 was printed :	90723 times
Number 6 was printed :	91053 times
Number 7 was printed :	90759 times
Number 8 was printed :	90872 times
Number 9 was printed :	91113 times
Number 10 was printed :	90637 times
--- Debugging process stopped ---
```

**Reply by Richt1707** [reply]
I'm very, very late to the party but, as others have pointed out, you are generating the random number within the `if` statements multiple times which means you are chasing a random number that you might never catch. I have taken the liberty of refactoring your code using an array to demonstrate an alternative way of doing this:

```Gdscript
extends Node2D

class RandomGeneration:
	func _generate_random_number():
		return randi_range(0,10)

func _ready() -> void:

	var loop_number = 1000000
	var number: Array[int] = []
	var idx: int
	var randomnumber: RandomGeneration = RandomGeneration.new()
	
	number.resize(11)
	
	while loop_number >= 0:
		
		idx = randomnumber._generate_random_number()
		
		number[idx] += 1

		if loop_number == 0:
			var idx2: int = 0
			var total: int = 0
			for num in number:
				print("Number %d was chosen %d times" % [idx2,num])
				total += num
				idx2 += 1
			
			print("Grand total %07d" % total)
			
		loop_number -= 1
```

This outputs:

```
Number 0 was printed 90624 times
Number 1 was printed 90968 times
Number 2 was printed 90736 times
Number 3 was printed 91047 times
Number 4 was printed 90413 times
Number 5 was printed 91061 times
Number 6 was printed 90961 times
Number 7 was printed 90785 times
Number 8 was printed 91095 times
Number 9 was printed 91280 times
Number 10 was printed 91031 times
Grand total 1000001
```
> [!NOTE]
> Your results will likely vary from mine as the above script was included within other scripts which may have already called other random functions during execution

**Comment by Mingu113** [comment]
There is no RandiRange, RandfRange in Godot C#?

**Reply by Viktorjb** [reply]
Struggled with this too! Appearently in C#, it's simply RandRange(). And it can take both integers and doubles, so for random floats I'm simply type casting double to float.
